pkg/statedb/reconciler/reconciler_test.go

```diff
@@ -0,0 +1,538 @@
+	h := testHelper{
+		t:      t,
+		db:     db,
+		tbl:    testObjects,
+		ops:    mt,
+		r:      r,
+		health: health,
+	}
+
+	numIterations := 10
+
+	t.Run("incremental", func(t *testing.T) {
+		h.t = t
+
+		for i := 0; i < numIterations; i++ {
+			t.Logf("Iteration %d", i)
+
+			// Insert some test objects and check that they're reconciled
+			t.Logf("Inserting test objects 1, 2 & 3")
+			h.insert(ID_1, NonFaulty, reconciler.StatusPending())
+			h.expectOp("update: 1")
+			h.expectStatus(ID_1, reconciler.StatusKindDone, "")
+
+			h.insert(ID_2, NonFaulty, reconciler.StatusPending())
+			h.expectOp("update: 2")
+			h.expectStatus(ID_2, reconciler.StatusKindDone, "")
+
+			h.insert(ID_3, NonFaulty, reconciler.StatusPending())
+			h.expectOp("update: 3")
+			h.expectStatus(ID_3, reconciler.StatusKindDone, "")
+
+			h.expectHealthLevel(cell.StatusOK)
+
+			// Set one to be faulty => object will error
+			t.Log("Setting '1' faulty")
+			h.insert(ID_1, Faulty, reconciler.StatusPending())
+			h.expectOp("update fail: 1")
+			h.expectStatus(ID_1, reconciler.StatusKindError, "update fail")
+			h.expectRetried(ID_1)
+			h.expectHealthLevel(cell.StatusDegraded)
+
+			// Fix the object => object will reconcile again.
+			t.Log("Setting '1' non-faulty")
+			h.insert(ID_1, NonFaulty, reconciler.StatusPending())
+			h.expectOp("update: 1")
+			h.expectStatus(ID_1, reconciler.StatusKindDone, "")
+			h.expectHealthLevel(cell.StatusOK)
+
+			t.Log("Delete 1 & 2")
+			h.markForDelete(ID_1)
+			h.expectOp("delete: 1")
+			h.expectNotFound(ID_1)
+
+			h.markForDelete(ID_2)
+			h.expectOp("delete: 2")
+			h.expectNotFound(ID_2)
+
+			t.Log("Try to delete '3' with faulty ops")
+			h.setTargetFaulty(true)
+			h.markForDelete(ID_3)
+			h.expectOp("delete fail: 3")
+			h.expectStatus(ID_3, reconciler.StatusKindError, "delete fail")
+			h.expectHealthLevel(cell.StatusDegraded)
+
+			t.Log("Delete 3")
+			h.setTargetFaulty(false)
+			h.expectOp("delete: 3")
+			h.expectNotFound(ID_3)
+			h.expectHealthLevel(cell.StatusOK)
+
+		}
+	})
+
+	t.Run("full", func(t *testing.T) {
+		h.t = t
+
+		for i := 0; i < numIterations; i++ {
+			t.Logf("Iteration %d", i)
+
+			// Without any objects, we should only see prune.
+			t.Log("Full reconciliation without objects")
+			h.triggerFullReconciliation()
+			h.expectOp("prune: 0")
+			h.expectHealthLevel(cell.StatusOK)
+
+			// Add few objects and wait until incremental reconciliation is done.
+			t.Log("Insert test objects")
+			h.insert(ID_1, NonFaulty, reconciler.StatusPending())
+			h.insert(ID_2, NonFaulty, reconciler.StatusPending())
+			h.insert(ID_3, NonFaulty, reconciler.StatusPending())
+			h.expectStatus(ID_1, reconciler.StatusKindDone, "")
+			h.expectStatus(ID_2, reconciler.StatusKindDone, "")
+			h.expectStatus(ID_3, reconciler.StatusKindDone, "")
+			h.expectHealthLevel(cell.StatusOK)
+
+			// Full reconciliation with functioning ops.
+			t.Log("Full reconciliation with non-faulty ops")
+			h.triggerFullReconciliation()
```

Comment:
maybe worth a comment _why_ we expect a full prune here
The test implementation of prune differs a bit from the expected implementation - the test indiscriminatorily deletes all, whereas a real prune should only delete unexpected, right?

Reply:
I'm not sure I follow. what's a full prune? we're just asserting that prune gets called with N objects which define the desired state and everything outside it is candidate for pruning.

Reply:
I tried improving the documentation for Prune(), hopefully clearer now.

Reply:
ah. well, I was misunderstanding the test DSL - I thought `prune: 3` means three things got pruned, not we called prune with three things. The latter of course makes more sense.